plugins/router/f5/plugin.go

```diff
@@ -0,0 +1,466 @@
+							member, poolName, err)
+					}
+				}
+			}
+		}
+		glog.V(4).Infof("Updated endpoints on F5.")
+	}
+
+	return nil
+}
+
+// HandleRoute processes watch events on the Route resource and
+// creates and deletes policy rules in response.
+func (p *F5Plugin) HandleRoute(eventType watch.EventType,
+	route *routeapi.Route) error {
+	glog.V(4).Infof("Processing route for service: %v (%v)",
+		route.ServiceName, route)
+
+	// In order to map OpenShift routes to F5 objects, we must divide routes into
+	// several types:
+	//
+	// • Insecure routes, those with no SSL/TLS, are implemented using a profile
+	//   on the HTTP vserver by creating a rule for each route.
+	//
+	// • "Secure" routes, comprising edge and reencrypt routes, are implemented
+	//   using a profile on the HTTPS vserver and rules on this profile, as well
+	//   as client SSL profiles and (for reencrypt) server SSL profiles.
+	//
+	// • Passthrough routes are implemented using an iRule that is associated with
+	//    the HTTPS vserver.  This iRule parses the SNI protocol and looks the
+	//    servername up in a datagroup to determine the pool for a request.  Thus
+	//    we must maintain a datagroup that maps hostname to poolname.
+	//
+	// As a result of this mapping from OpenShift concepts to F5 concepts, one may
+	// observe interesting behaviour when creating and modifying routes.  For
+	// example, creating a secure route, then creating a passthrough route with
+	// the same hostname, and then modifying the secure route will cause the
+	// passthrough route to be deleted.  Thus the passthrough route will not work
+	// even after the secure route is deleted.  This behaviour is seen as an
+	// acceptable trade-off between carrying a lot of extra state on our side and
+	// behaving a bit oddly when a user does silly things.
+
+	// Name of the pool in F5.
+	poolname := fmt.Sprintf("openshift_%s_%s",
+		route.Namespace, route.ServiceName)
+
+	// Virtual hostname for policy rule in F5.
+	hostname := route.Host
+
+	// Pathname for the policy rule in F5.
+	pathname := route.Path
+
+	// We will use prettyPathname for log output.
+	prettyPathname := pathname
+	if prettyPathname == "" {
+		prettyPathname = "(any)"
+	}
+
+	// Name for the route in F5.
+	routename := fmt.Sprintf("openshift_route_%s_%s", route.Namespace, route.Name)
+
+	modifying := false
+
+	switch eventType {
+	case watch.Modified:
+		glog.V(4).Infof("Updating route %s...", routename)
+
+		// We handle modifications by deleting the old route (by falling through
+		// here) and then adding the updated route (by falling through from the
+		// watch.Deleted case to the watch.Added case).
+		modifying = true
+		fallthrough
+
+	case watch.Deleted:
+		glog.V(4).Infof("Deleting route %s...", routename)
+
+		// Start with the routes because we cannot delete the pool until we delete
+		// any associated profiles and rules.
+
+		secureRouteExists, err := p.F5Client.SecureRouteExists(routename)
```

Comment:
Do we need to make these calls every time for every route or should we base this (and the call below) on the TLS section of the route?  I imagine it doesn't save a whole lot unless there are lots of updates going on.